packages/plugin-npm-cli/sources/npmAuditUtils.ts

```diff
@@ -206,51 +87,60 @@ export function getRequires(project: Project, workspace: Workspace, {all, enviro
   ].includes(environment);
 
-  const requiredDevDependencies = [];
-  if (includeDevDependencies)
-    for (const workspace of workspaces)
-      for (const dependency of workspace.manifest.devDependencies.values())
-        requiredDevDependencies.push(dependency);
+  for (const workspace of workspaces) {
+    for (const dependency of workspace.anchoredPackage.dependencies.values()) {
+      const isDevDependency = workspace.manifest.devDependencies.has(dependency.identHash);
+      if (isDevDependency ? !includeDevDependencies : !includeDependencies)
+        continue;
 
-  return transformDescriptorIterableToRequiresObject([
-    ...requiredDependencies,
-    ...requiredDevDependencies,
-  ].filter(dependency => {
-    return structUtils.parseRange(dependency.range).protocol === null;
-  }));
+      topLevelDependencies.push({workspace, dependency});
+    }
+  }
+
+  return topLevelDependencies;
 }
 
-export function getDependencies(project: Project, workspace: Workspace, {all}: {all: boolean}) {
-  const transitiveDevDependencies = getTransitiveDevDependencies(project, workspace, {all});
+export function getPackages(project: Project, roots: Array<TopLevelDependency>, {recursive}: {recursive: boolean}) {
+  const packages = new Map<string, Map<string, Array<Locator>>>();
+  const traversed = new Set<LocatorHash>();
 
-  const data: {
-    [key: string]: {
-      version: string;
-      integrity: string;
-      requires: {[key: string]: string};
-      dev: boolean;
-    };
-  } = {};
+  const processDescriptor = (parent: Locator, descriptor: Descriptor) => {
+    const resolution = project.storedResolutions.get(descriptor.descriptorHash);
+    if (typeof resolution === `undefined`)
+      throw new Error(`Assertion failed: The resolution should have been registered`);
 
-  for (const pkg of project.storedPackages.values()) {
-    data[structUtils.stringifyIdent(pkg)] = {
-      version: pkg.version ?? `0.0.0`,
-      integrity: pkg.identHash,
-      requires: transformDescriptorIterableToRequiresObject(pkg.dependencies.values()),
-      dev: transitiveDevDependencies.has(structUtils.convertLocatorToDescriptor(pkg).descriptorHash),
-    };
-  }
+    if (!traversed.has(resolution))
+      traversed.add(resolution);
+    else
+      return;
+
+    const pkg = project.storedPackages.get(resolution);
+    if (typeof pkg === `undefined`)
+      throw new Error(`Assertion failed: The package should have been registered`);
+
+    const devirtualizedLocator = structUtils.ensureDevirtualizedLocator(pkg);
+
+    if (devirtualizedLocator.reference.startsWith(`npm:`) && pkg.version !== null) {
+      const packageName = structUtils.stringifyIdent(pkg);
+
+      const versions = miscUtils.getMapWithDefault(packages, packageName);
+      miscUtils.getArrayWithDefault(versions, pkg.version).push(parent);
+    }
+
+    if (recursive) {
+      for (const dependency of pkg.dependencies.values()) {
+        processDescriptor(pkg, dependency);
+      }
+    }
```

Comment:
I'm a bit worried about this causing a `Maximum call stack size exceeded` error, could you make it iterative instead?